shell/src/main/java/alluxio/cli/fs/command/PersistCommand.java

```diff
@@ -47,71 +68,116 @@ public String getCommandName() {
+    }
     String[] args = cl.getArgs();
+    List<AlluxioURI> candidateUris = new ArrayList<>();
     for (String path : args) {
-      AlluxioURI inputPath = new AlluxioURI(path);
-      runWildCardCmd(inputPath, cl);
+      candidateUris.addAll(FileSystemShellUtils.getAlluxioURIs(mFileSystem, new AlluxioURI(path)));
+    }
+    final BlockingQueue<AlluxioURI> toPersist = new LinkedBlockingQueue<>();
+    for (AlluxioURI uri : candidateUris) {
+      queuePersist(mFileSystem.getStatus(uri), toPersist);
+    }
+    System.out.println("Found " + toPersist.size() + " files to persist.");
+    ExecutorService service =
+        Executors.newFixedThreadPool(parallelism, ThreadFactoryUtils.build("persist-cli-%d", true));
+    final Object progressLock = new Object();
+    AtomicInteger completedFiles = new AtomicInteger(0);
+    List<Future<Void>> futures = new ArrayList<>(parallelism);
+    for (int i = 0; i < parallelism; i++) {
+      futures.add(service.submit(new PersistCallable(toPersist, completedFiles, progressLock)));
+    }
+    try {
+      for (Future<Void> future : futures) {
+        future.get();
+      }
+    } catch (ExecutionException e) {
+      System.out.println("Fatal error: " + e);
+      return -1;
+    } catch (InterruptedException e) {
+      System.out.println("Persist interrupted, exiting.");
+      return -1;
+    } finally {
+      service.shutdownNow();
     }
     return 0;
   }
 
+  private void queuePersist(URIStatus status, BlockingQueue<AlluxioURI> toPersist)
+      throws AlluxioException, IOException {
+    AlluxioURI uri = new AlluxioURI(status.getPath());
+    if (status.isFolder()) {
+      List<URIStatus> statuses = mFileSystem.listStatus(uri);
+      for (URIStatus s : statuses) {
+        queuePersist(s, toPersist);
+      }
+    } else if (!status.isPersisted()) {
+      toPersist.add(uri);
+    }
+  }
+
   /**
-   * Persists a file or directory currently stored only in Alluxio to the UnderFileSystem.
-   *
-   * @param filePath the {@link AlluxioURI} path to persist to the UnderFileSystem
+   * Thread that polls a persist queue and persists files.
    */
-  private void persist(AlluxioURI filePath) throws AlluxioException, IOException {
-    URIStatus status = mFileSystem.getStatus(filePath);
-    if (status.isFolder()) {
-      List<URIStatus> statuses = mFileSystem.listStatus(filePath);
-      List<String> errorMessages = new ArrayList<>();
-      for (URIStatus uriStatus : statuses) {
-        AlluxioURI newPath = new AlluxioURI(uriStatus.getPath());
+  private class PersistCallable implements Callable<Void> {
+    private final BlockingQueue<AlluxioURI> mToPersist;
+    private final int mTotalFiles;
+    private final Object mProgressLock;
+    private final AtomicInteger mCompletedFiles;
+
+    PersistCallable(BlockingQueue<AlluxioURI> toPersist, AtomicInteger count, Object progressLock) {
+      mToPersist = toPersist;
+      mTotalFiles = toPersist.size();
```

Comment:
can we pass the total instead so that all threads get the initial `toPersist` size? Otherwise the 1st thread could have reduced the size of `toPersist` by the time 10th thread hits this line

Reply:
Yeah I just cut a corner here because the way it is currently implemented we init all the `PersistCallables` (serial) before we submit.

Reply:
Are you sure we init the `PersistCallables` before running them? It doesn't look like it here: `futures.add(service.submit(new PersistCallable(toPersist, completedFiles, progressLock)));`

But even if we did, it's a subtle dependency on the usage that could be easily avoided, so I still think it's better to defensively take `totalFiles` as an arg